luigi_pipeline/lib/model/base_mt_schema.py

```diff
@@ -0,0 +1,105 @@
+from abc import abstractmethod
+from inspect import getmembers, isfunction
+from functools import wraps
+
+
+def mt_annotation(annotation=None, fn_require=None):
+    """
+    Function decorator for methods in a subclass of BaseMTSchema.
+    Allows the function to be treated like an mt_annotation with annotation name and value.
+
+        @mt_annotation()
+        def a(self):
+            return 'a_val'
+
+        @mt_annotation(annotation='b', fn_require=a)
+        def b_1(self):
+            return 'b_val'
+
+    Will generate a mt with rows of {a: 'a_val', 'b': 'b_val'} if the function is called.
+    TODO: Consider changing fn_require to be a list of requirements.
+
+    :param annotation: name in the final MT. If not provided, uses the function name.
+    :param fn_require: method name strings in class that are dependencies.
+    :return:
+    """
+    def mt_prop_wrapper(func):
+        annotation_name = annotation or func.__name__
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # Called already.
+            if wrapper.mt_prop_meta['annotated'] > 0:
+                return self
+            if fn_require:
+                getattr(self, fn_require.__name__)()
+
+            # Annotate and retiurn instance for chaining.
+            self.mt = self.mt.annotate_rows(**{annotation_name: func(self, *args, **kwargs)})
+            wrapper.mt_prop_meta['annotated'] += 1
+            return self
+
+        wrapper.mt_prop_meta = {
+            'annotated': 0,  # Counter for number of times called. Should only be 0 or 1.
+            'annotated_name': annotation_name
+        }
+        return wrapper
+    return mt_prop_wrapper
+
+
+class BaseMTSchema:
+    """
+    Main superclass that provides a Hail MT schema definition. decorate methods with @mt_annotation.
+    Allows annotations to express dependencies where dependencies are run before (and at most once).
+    NOTE: circular dependencies are not supported and not gracefully handled.
+
+    Usage example:
+        class TestSchema(BaseMTSchema):
+
+            def __init__(self):
+                super(TestSchema, self).__init__(hl.import_vcf('tests/data/1kg_30variants.vcf.bgz'))
+
+            @mt_annotation()
+            def a(self):
+                return 0
+
+            @mt_annotation(fn_require=a)
+            def b(self):
+                return self.a + 1
+
+            @mt_annotation(annotation='c', fn_require=a)
+            def c_1(self):
+                return self.a + 2
+
+    `TestSchema(mt).b().c_1().select_annotated_mt()` will annotate with {'a': 0, 'b': 1, 'c': 2}
```

Comment:
This API seems a bit unwieldy. To me, `Schema(mt).foo()` looks like it would return an expression for the `foo` annotation vs modifying the instance's internal MatrixTable.

It also seems unlikely that someone creating a schema would ever want some but not all of the annotation fields, so the common use case would be `TestSchema(mt).annotate_all().mt`? Would something like `TestSchema.annotate_mt(mt)` work?

Reply:
So I thought it might be confusing, but thought that the `Schema` class would make it more obvious that it's adding annotations. Even thought of naming it `AnnotationBuilder` but `Schema` seems more appropriate since it'll do more.

So adding this functionality just provides more flexibility. (e.g. it was really helpful in our test cases, when we just wanted to test derived fields). It also allows the client flexibility if they wanted partials for some reason (so I was thinking if we wanted to, we can explicitly add stages for derived fields, reference data, etc. Let's provide the functionality for now, and remove (or not advertise) if it's too confusing.